### PR TITLE
Refine soma_ column handling in Python API

### LINKED ISSUE #324: Python tiledbsoma feature: create accepts soma_rowid in schema
For classes such as SOMADataFrame, there is an inconvenience in the create operator handling of the `soma_rowid` field in the schema:
* create will not accept a `soma_rowid` column, even if it is correctly typed as uint64
* read returns record batches / tables with this field present in the arrow result schema
* get_schema will return a schema that includes `soma_rowid`

The current abstract spec suggests (in NdArray) that get_schema should return the schema of data returned by the read operation.  I think it would be very convenient if `create` was slightly more flexible in its handling:
* if `soma_rowid` is present, just confirms that it is correctly typed uint64, errors out if not
* if the field is not present, adds it silently

### LINKED ISSUE #325: Python tiledbsoma: soma_joinid missing
The abstract spec says:
> `SOMADataFrame` must contain a column called `soma_joinid`, of type `uint64`. The `soma_joinid` column contains a unique value for each row in the `SOMADataFrame`, and intended to act as a joint key for other objects, such as `SOMASparseNdArray`.

and
>All `SOMAIndexedDataFrame` must contain a column called `soma_joinid`, of type `uint64`. The `soma_joinid` column contains a unique value for each row in the `SOMAIndexedDataFrame`, and intended to act as a joint key for other objects, such as `SOMASparseNdArray`.

However, the current Python implementation does not implement this.

### LINKED ISSUE #329: Py tiledbsoma SOMAIndexedDataFrame read in pandas format not implemented
the "read" operations into pandas DataFrame needs to be implemented

### PR DESCRIPTION
**Pull request by bkmartinjr**
Update the Python API to match latest changes to the SOMA abstract spec around:
* NdArray dim/data column names
* Index data type (now int64)
* Explicit handling of soma_joinid/soma_rowid

Changes:
* soma_joinid and soma_rowid change from uint64 to int64 in Python and C++
* NDArray dimensions now named soma_dim_N; data column now named soma_data
* DataFrame/IndexDataFrame: create will now enforce soma_rowid/soma_joinid typing, and automatically create as needed
* `soma.io.from_anndata()` no longer uses TileDB-Py.from_pandas() code path, but instead uses the core SOMA API to write, allowing a single code path to implement all schema-related validation, defaults, etc. To allow for gentle transition, added behind a feature flag (`from_anndata_write_pandas_using_arrow`)
* DataFrame:  soma.io-specific code moved into io.py
* Unit tests related to the above changes

Minor changes:
* rewrite DataFrame/IndexedDataFrame read_as_pandas to use underlying Arrow read
* add new constants.py file containing commonly used constant values
* revise keys() to match spec:  return all schema field names, not just TileDB attributes
* remove shape/ndims from dataframe classes as they are not part of the spec

Side effects of this PR:
* Fixes #329 
* Python package implementation of #384 
* Fixes #325 
* Fixes #324 
* Python package implementation of #396

### CODE REVIEW

File: apis/python/tests/test_soma_dataframe.py [hunk 1]
```diff
@@ -18,11 +18,17 @@ def test_soma_dataframe_non_indexed(tmp_path):
...
+        ["foo", "bar", "baz", "soma_rowid", "soma_joinid"]
+    )
+    assert sorted(sdf.keys()) == sorted(sdf.schema.names)
+
     # ----------------------------------------------------------------
     # Write
     for _i in range(3):
         pydict = {}
         pydict["soma_rowid"] = [0, 1, 2, 3, 4]
+        pydict["soma_joinid"] = [0, 1, 2, 3, 4]
```
Comment (johnkerl) [inline apis/python/tests/test_soma_dataframe.py:31]: can't hurt to make these `[100,400,500,200,300]` just to prove the point that they're truly arbitrary ...

File: libtiledbsoma/src/pyapi/libtiledbsoma.cc [hunk 2]
```diff
@@ -205,26 +198,13 @@ PYBIND11_MODULE(libtiledbsoma, m) {
                     array.attr("_export_to_c")(
                         arrow_array_ptr, arrow_schema_ptr);
 
-                    if (!strcmp(arrow_schema.format, "L") ||
-                        !strcmp(arrow_schema.format, "l")) {
-                        std::span<uint64_t> data{
-                            (uint64_t*)arrow_array.buffers[1],
+                    // 'l' = arrow data type int64
```
Comment (johnkerl) [inline libtiledbsoma/src/pyapi/libtiledbsoma.cc:201]: thank you for this comment ❤️

**Review by johnkerl — APPROVED** [review]
LGTM

this is a big PR, but, everything reads reasonably & straightforwardly & feels like it just "should be this way"

suggestion to make join IDs not be 0 .. n-1 but rather something else (at all callsites)

i'm approving, not requesting changes, as i know you can make that change & i want you to be able to merge this whenever especially with regard to timezones

thanks @bkmartinjr !! :)

File: apis/python/tests/test_soma_dataframe.py (hunk 1 above)
Comment (bkmartinjr) [reply apis/python/tests/test_soma_dataframe.py]: Good idea - fixed.

File: libtiledbsoma/src/pyapi/libtiledbsoma.cc (hunk 2 above)
Comment (bkmartinjr) [reply libtiledbsoma/src/pyapi/libtiledbsoma.cc:201]: For reference, see [Arrow C inteface format strings](https://arrow.apache.org/docs/format/CDataInterface.html#data-type-description-format-strings)

**Review by thetorpedodog — COMMENTED** [review]
I think most of the stuff I have here is fairly minor; the overall direction here is good.

File: apis/python/src/tiledbsoma/soma_indexed_dataframe.py [hunk 3]
```diff
@@ -47,35 +45,65 @@ def type(self) -> Literal["SOMAIndexedDataFrame"]:
...
+        if SOMA_JOINID in schema.names:
+            if schema.field(SOMA_JOINID).type != pa.int64():
+                raise TypeError(f"{SOMA_JOINID} field must be of type Arrow int64")
+        else:
+            # add SOMA_JOINID
+            schema = schema.append(pa.field(SOMA_JOINID, pa.int64()))
+
+        # verify no illegal use of soma_ prefix
+        for field_name in schema.names:
+            if field_name.startswith("soma_") and field_name not in [SOMA_JOINID]:
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/soma_indexed_dataframe.py]: fixed

File: apis/python/src/tiledbsoma/soma_indexed_dataframe.py [hunk 4]
```diff
@@ -47,35 +45,65 @@ def type(self) -> Literal["SOMAIndexedDataFrame"]:
...
 
+    def _validate_schema(
+        self, schema: pa.Schema, index_column_names: Sequence[str]
+    ) -> pa.Schema:
+        """
+        Handle default column additions (eg, soma_joinid) and error checking on required columns.
+
+        Returns a schema, which may be modified by the addition of required columns.
+        """
+        if index_column_names is None or len(index_column_names) == 0:
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/soma_indexed_dataframe.py]: fixed

File: apis/python/src/tiledbsoma/soma_indexed_dataframe.py [hunk 5]
```diff
@@ -47,35 +45,65 @@ def type(self) -> Literal["SOMAIndexedDataFrame"]:
...
-            assert index_column_name in schema_names_set
-
-        assert ROWID not in index_column_names
-        assert ROWID not in schema_names_set
-
+        schema = self._validate_schema(schema, index_column_names)
         self._create_empty(schema, index_column_names)
         self._is_indexed = True
-        self._index_column_names = index_column_names
+        self._index_column_names = list(index_column_names)
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/soma_indexed_dataframe.py]: fixed. Pulling this thread triggered a bunch of other type signature changes, all of which are improvements IMHO.

File: apis/python/src/tiledbsoma/soma_dataframe.py [hunk 6]
```diff
@@ -256,19 +267,25 @@ def write(self, values: pa.Table) -> None:
...
 
-        The ``values`` Arrow Table must contain a ``soma_rowid`` (uint64) column, indicating which rows are being written.
+        The ``values`` Arrow Table must contain a ``soma_rowid`` (int64) column, indicating which rows are being written.
         """
-        self._shape = None  # cache-invalidate
+        if (
+            SOMA_ROWID not in values.schema.names
+            or SOMA_JOINID not in values.schema.names
+        ):
+            raise TypeError(
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/soma_dataframe.py]: agree.  fixed.

File: apis/python/src/tiledbsoma/soma_dataframe.py [hunk 7]
```diff
@@ -256,19 +267,25 @@ def write(self, values: pa.Table) -> None:
...
+                f"Write requires both {SOMA_ROWID} and {SOMA_JOINID} to be specified in table."
+            )
 
-        assert ROWID in values.schema.names
+        assert SOMA_ROWID in values.schema.names
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/soma_dataframe.py]: fixed

File: apis/python/src/tiledbsoma/soma_dataframe.py [hunk 8]
```diff
@@ -51,31 +48,61 @@ def create(
...
-
+        schema = self._validate_schema(schema)
         self._create_empty(schema)
         self._is_indexed = False
         self._index_column_names = []
 
         self._common_create()  # object-type metadata etc
         return self
 
+    def _validate_schema(self, schema: pa.Schema) -> pa.Schema:
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/soma_dataframe.py]: yes, makes good sense.  also fixed it in soma_indexed_dataframe.py

File: apis/python/src/tiledbsoma/io.py [hunk 9]
```diff
@@ -82,6 +78,75 @@ def _from_h5ad_common(
...
+        )
+        dim_filters = tiledb.FilterList([tiledb.ZstdFilter(level=-1)])
+        attr_filters = tiledb.FilterList([tiledb.ZstdFilter(level=-1)])
+
+        # Force ASCII storage if string, in order to make obs/var columns queryable.
+        # TODO: when UTF-8 attributes are fully supported we can remove this.
+        column_types = {}
+        for column_name in df.keys():
+            dfc = df[column_name]
+            if len(dfc) > 0 and type(dfc[0]) == str:
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/io.py]: fixed

**Comment by johnkerl** [comment]
🚀

**Pull request merged**